src/db_reader.rs

```diff
@@ -0,0 +1,1083 @@
+            lifetime: Some(self.options.checkpoint_lifetime),
+            ..CheckpointOptions::default()
+        };
+        stored_manifest
+            .replace_checkpoint(current_checkpoint_id, &options)
+            .await
+    }
+
+    async fn reestablish_checkpoint(&self, checkpoint: Checkpoint) -> Result<(), SlateDBError> {
+        let new_checkpoint_state = self.rebuild_checkpoint_state(checkpoint).await?;
+        let mut write_guard = self.state.write();
+        *write_guard = new_checkpoint_state;
+        Ok(())
+    }
+
+    async fn maybe_replay_new_wals(&self) -> Result<(), SlateDBError> {
+        let last_seen_wal_id = self.table_store.last_seen_wal_id().await?;
+        let last_replayed_wal_id = self.state.read().last_wal_id;
+        if last_seen_wal_id > last_replayed_wal_id {
+            let mut updated_checkpoint = self.state.read().clone();
+            let mut updated_state = updated_checkpoint.state.as_ref().clone();
+
+            Self::replay_wal_into(
+                Arc::clone(&self.table_store),
+                &self.options,
+                updated_checkpoint.state.core(),
+                &mut updated_state.imm_memtable,
+                true,
+            )
+            .await?;
+            updated_checkpoint.state = Arc::new(updated_state);
+            let mut write_guard = self.state.write();
+            *write_guard = updated_checkpoint;
+        }
+        Ok(())
+    }
+
+    async fn build_initial_checkpoint_state(
+        manifest_store: Arc<ManifestStore>,
+        table_store: Arc<TableStore>,
+        options: &DbReaderOptions,
+        checkpoint: Checkpoint,
+        replay_new_wals: bool,
+    ) -> Result<CheckpointState, SlateDBError> {
+        let manifest = manifest_store.read_manifest(checkpoint.manifest_id).await?;
+        let mut imm_memtable = VecDeque::new();
+        let last_wal_id = Self::replay_wal_into(
+            Arc::clone(&table_store),
+            options,
+            &manifest.core,
+            &mut imm_memtable,
+            replay_new_wals,
+        )
+        .await?;
+
+        let dirty_manifest = DirtyManifest::new(checkpoint.manifest_id, manifest);
+
+        let state = Arc::new(COWDbState {
+            imm_memtable,
+            imm_wal: VecDeque::new(),
+            manifest: dirty_manifest,
+        });
+
+        Ok(CheckpointState {
+            checkpoint,
+            state,
+            last_wal_id,
+        })
+    }
+
+    async fn rebuild_checkpoint_state(
+        &self,
+        new_checkpoint: Checkpoint,
+    ) -> Result<CheckpointState, SlateDBError> {
+        let prior = self.state.read().clone();
+        let manifest = self
+            .manifest_store
+            .read_manifest(new_checkpoint.manifest_id)
+            .await?;
+
+        let mut imm_memtable = prior
+            .state
+            .imm_memtable
+            .iter()
+            .filter(|table| table.last_wal_id() <= manifest.core.last_compacted_wal_sst_id)
```

Comment:
I think this will potentially result in some overlap between the imm_memtables and l0, which is fine because `get` and the merge iterators will just deal with that naturally. But maybe leave a comment?